File: src/linear.zig
const std = @import("std");

const matrix_vector_product = @import("simd/matrix_vector_product.zig");
const rms_norm = @import("simd/rms_norm.zig");
const softmax = @import("simd/softmax.zig");

const Self = @This();

norm_weight: []f32,
output_weight: []f32,
input_vector: []f32,
output_vector: []f32,

// private
norm_vector: []f32,

export fn create(input_size: usize, output_size: usize) ?*Self {
    const allocator = std.heap.page_allocator;
    const self = allocator.create(Self) catch return null;

    self.* = .{
        .norm_weight = allocator.alloc(f32, input_size) catch return null,
        .output_weight = allocator.alloc(f32, output_size * input_size) catch return null,
        .input_vector = allocator.alloc(f32, input_size) catch return null,
        .output_vector = allocator.alloc(f32, output_size) catch return null,
        .norm_vector = allocator.alloc(f32, input_size) catch return null,
    };

    return self;
}

export fn getNormWeight(self: *const Self) [*]f32 {
    return self.norm_weight.ptr;
}

export fn getOutputWeight(self: *const Self) [*]f32 {
    return self.output_weight.ptr;
}

export fn getInputVector(self: *const Self) [*]f32 {
    return self.input_vector.ptr;
}

export fn getOutputVector(self: *const Self) [*]f32 {
    return self.output_vector.ptr;
}

export fn forward(self: *const Self, compute_softmax: bool) void {
    rms_norm.compute(self.norm_weight, self.input_vector, self.norm_vector);
    matrix_vector_product.compute(self.output_weight, self.norm_vector, self.output_vector);

    if (compute_softmax) {
        softmax.compute(self.output_vector, self.output_vector);
    }
}

File: packages/decoder/src/linear.ts
import { checkNotNull } from './check_not_null.js';

export interface LinearInit {
  readonly inputSize: number;
  readonly outputSize: number;
}

export class Linear {
  static wasmSingleton: WebAssembly.Instance;

  readonly normWeight: Uint8Array;
  readonly outputWeight: Uint8Array;
  readonly inputVector: Float32Array;
  readonly outputVector: Float32Array;

  private readonly self: number;

  constructor({ inputSize, outputSize }: LinearInit) {
    const { memory, create, getNormWeight, getOutputWeight, getInputVector, getOutputVector } =
      Linear.wasmSingleton.exports as WasmExports;

    this.self = create(inputSize, outputSize);

    this.normWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getNormWeight(this.self)),
      inputSize * 4,
    );

    this.outputWeight = new Uint8Array(
      memory.buffer,
      checkNotNull(getOutputWeight(this.self)),
      outputSize * inputSize * 4,
    );

    this.inputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getInputVector(this.self)),
      inputSize,
    );

    this.outputVector = new Float32Array(
      memory.buffer,
      checkNotNull(getOutputVector(this.self)),
      outputSize,
    );
  }

  forward(): void {
    const { forward } = Linear.wasmSingleton.exports as WasmExports;

    forward(this.self);
  }

  computeSoftmax(): void {
    const { computeSoftmax } = Linear.wasmSingleton.exports as WasmExports;

    computeSoftmax(this.self);
  }
}

interface WasmExports extends WebAssembly.Exports {
  readonly memory: WebAssembly.Memory;

  create(inputSize: number, outputSize: number): number;
  getNormWeight(self: number): number;
  getOutputWeight(self: number): number;
  getInputVector(self: number): number;
  getOutputVector(self: number): number;
  forward(self: number): void;
  computeSoftmax(self: number): void;
}
